### ISSUE TITLE
About Modifying the Wave Equation Support

### ISSUE THREAD

**Issue by StarrMoonn** [ar4/deepwave#96]
Hello author, I want to modify the wave equation, which files should I modify? Because pure Pytorch's FD and gradient calculations are too slow, I am considering porting VTI's code. I have pure C and Cuda codes on hand.

There is another question. In traditional FWI code, after proposing an error function, many articles will draw and analyze the "companion source". However, our Pytorch-based deepwave, especially after I modify the error function, this is automatic differentiation to achieve grafting. There is a direct "bridge" with the gradient calculation of cuda, right? Can I still draw the "companion source" of this new error function?

Because I see that FD and gradient are both handwritten operators, but we can customize the misfit function. This should be pytorch to achieve the bridge, right? But there is a problem here. The traditional gradient calculation adjoint method, AD is also based on the chain rule principle, but this adjoint method is the waveform derivative. For strong scattering such as the salt dome model of EAGE, it seems that this gradient calculation does not work. I see that they are based on the energy method Born approximation to deduce the adjoint wave equation, so I wonder if I can support the injection of handwritten adjoint sources into deepwave? Fully utilize the powerful calculation speed of deepwave's FD and backpropagation!

**Comment by ar4** [comment]
Hello @StarrMoonn 

Thank you for your questions. I am not certain that I understand all of them, so please let me know if I have not answered any of your questions.

> which files should I modify?

It depends on what you wish to do. It sounds like you want to implement a new propagator for VTI propagation. In that case, I suggest that you take one of the existing propagators as a base, such as the scalar propagator, copy its files, and modify them. For the scalar propagator, these files are scalar.c (for the CPU implementation), scalar.cu (for the GPU implementation), and scalar.py (the Python that acts as the bridge between PyTorch and the C/CUDA files). You will need to start by [downloading all of the Deepwave files](https://github.com/ar4/deepwave/archive/refs/heads/master.zip). You will find the scalar files in the `src/deepwave` directory. I suggest that you rename the copies you make by replacing `scalar` with `vti`. Once you have finished your implementation, you will then need to make two more changes. The first is to adjust the build script so that your new C/CUDA files are compiled. To do this, you should edit the file `src/deepwave/build_<your operating system>.sh`. You should add entries for your new files. You should then run the build script to generate the compiled library files. The other file you will need to modify is `src/deepwave/__init__.py`. This is where the compiled libraries are loaded into Python. You should add entries for your new functions.

> In traditional FWI code, after proposing an error function, many articles will draw and analyze the "companion source". However, our Pytorch-based deepwave, especially after I modify the error function, this is automatic differentiation to achieve grafting. There is a direct "bridge" with the gradient calculation of cuda, right? Can I still draw the "companion source" of this new error function?

I think you are referring to what I know as the "adjoint source". The adjoint source will be automatically created by PyTorch based on the operations that you apply to the output receiver data from Deepwave, such as what error/loss/objective function you use. PyTorch passes this adjoint source to Deepwave as the gradient with respect to the receiver data, and Deepwave then injects it into the backward propagation at the receiver locations. So, Deepwave has no involvement in the calculation of the adjoint source - PyTorch does that automatically and passes it to Deepwave. I am not sure what you mean by "draw" it, but I think you mean obtain it so that you can analyse it. I think the easiest way to do this is to [create a PyTorch backward hook](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html). If you create a backward hook for the receiver data, then I think it should pass the adjoint source into the function you define in the hook, which you can then use to analyse it, save it for later analysis, or modify it before it gets passed into Deepwave for backpropagation.

I don't think I understand what you are saying in your final paragraph, but it sounds like you wish to modify the adjoint source that gets used by Deepwave. If so, then I hope my answer in the previous paragraph will enable you to do what you wish, however I suggest that the best way to modify the adjoint source is to modify your error/loss/objective function.
